### ISSUE TITLE
Wayland: `kitty` crashes with `Broken pipe` after a lot of CLI output

### ISSUE THREAD

**Issue by Ma27** [kovidgoyal/kitty#2329]
I've seen that there are some similar bugreports[1], so I'm sorry if that's actually a duplicate, unfortunately I'm not too experienced with Wayland, so I may be missing something obvious here.

I observed that `kitty`-windows crash with `[glfw error 65544]: Wayland: fatal display error: Broken pipe` when I run a command that produces a lot of output and I leave kitty in a different workspace for a longer time (when I switch to that workspace occasionally and focus the kitty-window, it doesn't crash).

Please note that despite the crash there aren't any records in `coredumpctl(1)`.

I recorded the output of `WAYLAND_DEBUG=1 ./result/bin/kitty 2>&1 | tee log-wl-kitty.txt` in a GitHub Gist: https://gist.githubusercontent.com/Ma27/4190b3fd56e664c024ab8ffe0f44192b/raw/315bbafc999f80a92651bc1510dd817f7bb234f0/log.txt

Is there anything else I can do to investigate this issue?

---

OS: NixOS 19.09
kitty: 0.15.1 (also reproducible with 0.16.0), no additional config needed to reproduce
WM: sway 1.2
Mesa 19.1.5

---

[1] https://github.com/kovidgoyal/kitty/issues/1760, https://github.com/kovidgoyal/kitty/issues/1722

**Comment by ctrlcctrlv** [comment]
It might help to get a traceback. Run kitty in GDB. Be sure to compile with CFLAGS `-g -O0`, then `gdb -ex run path/to/compiled/kitty`

**Comment by kovidgoyal** [comment]
I'm going to need atleast a backtrace and ideally a set of steps to
reproduce.

make debug

then produce the crash and

echo bt | coredumpctl gdb

**Comment by kovidgoyal** [comment]
no followup

**Comment by akvadrako** [comment]
I'm not sure if I should make a new issue or comment here, but I have been having this problem for a while so I went and made a coredump.

```
#0  0x00007efe72680a66 in __ppoll (fds=0x7efe7074eae0 <_glfw+132448>, fds@entry=0x28, nfds=2, 
    timeout=<optimized out>, timeout@entry=0x7ffc21716410, sigmask=sigmask@entry=0x0)
    at ../sysdeps/unix/sysv/linux/ppoll.c:44
#1  0x00007efe70701e63 in ppoll (__ss=0x0, __timeout=0x7ffc21716410, __nfds=<optimized out>, __fds=0x28)
    at /usr/include/x86_64-linux-gnu/bits/poll2.h:77
#2  pollWithTimeout (fds=fds@entry=0x7efe7074eae0 <_glfw+132448>, nfds=<optimized out>, 
    timeout=timeout@entry=2978986) at glfw/backend_utils.c:177
#3  0x00007efe707021c8 in pollForEvents (eld=0x7efe7074eae0 <_glfw+132448>, timeout=2978986, 
    timeout@entry=-1, display_callback=display_callback@entry=0x7efe706f8f3e <wayland_read_events>)
    at glfw/backend_utils.c:306
#4  0x00007efe706f8ee5 in handleEvents (timeout=-1) at glfw/wl_window.c:793
#5  0x00007efe706fa579 in _glfwPlatformWaitEvents () at glfw/wl_window.c:1251
#6  0x00007efe706f72e4 in _glfwPlatformRunMainLoop (tick_callback=0x7efe713db214 <process_global_state>, 
    data=0x7efe70b39490) at glfw/main_loop.h:30
#7  0x00007efe706ef0a2 in glfwRunMainLoop (callback=<optimized out>, data=<optimized out>) at glfw/init.c:344
#8  0x00007efe714063d4 in run_main_loop (cb=cb@entry=0x7efe713db214 <process_global_state>, 
    cb_data=cb_data@entry=0x7efe70b39490) at kitty/glfw.c:1153
#9  0x00007efe713d8c47 in main_loop (self=0x7efe70b39490, a=<optimized out>) at kitty/child-monitor.c:975
```

**Comment by akvadrako** [comment]
This is on latest Ubuntu with sway 1.4-dffc184a

**Comment by kovidgoyal** [comment]
there's no crash in your backtrace. The last call is to the libc function ppoll, I'd be amazed if that crashed.

**Comment by akvadrako** [comment]
Ah, you're right sorry, I got them mixed up - I'll have one soon.

**Comment by akvadrako** [comment]
Are you sure it'll make a coredump when it encounters this error?

**Comment by kovidgoyal** [comment]
I have no idea, the OP said it was a crash. You should maybe cross post this on sway, a broken pipe typically means the compositor broke the connection. If sway has some limit on how much input it can handle, they would be the best people to ask.

**Comment by akvadrako** [comment]
I would call this a crash, though perhaps from you would prefer the term aborts.

I'll ask the sway team.

**Comment by akvadrako** [comment]
From the initial discussion over on swaywm it's hard to tell if it's a sway bug. But it surely seems like undesirable behavior in kitty. If I'm reading the timestamps correctly, kitty spends 11 seconds doing nothing but sending `delete_id` calls before it crashes.

**Comment by martinetd** [comment]
(The `delete_id` comes from sway, but kitty spent just this long creating this many frame callback requests that sway piled up and had to clean up before doing any useful work)

**Comment by kovidgoyal** [comment]
kitty doesn't send delete_id and it sends at most one request for a frame callback every quarter second. Are you saying sway really cannot handle that rate of requests?? If sway has no intention of honoring the render request, it should drop the request immediately. I dont see why it thinks piling them up for unbounded amounts of time is a good idea. kitty will request a render frame whenever something changes causing it to need to re-render, rate-limited to 4 per second precisely so as to not overwhelm the compositor. I could of course change that rate to 1 per infinite time but that requires trusting that all compositors will never drop a render frame request, no matter what. That is not a robust design. sway should learn to handle multiple render frame requests sent at a reasonable rate.

**Comment by akvadrako** [comment]
I’m know very little about the wayland protocol so I don’t think it makes sense to act as the go between. Wouldn’t it make more sense to put that comment in the sway bug?

**Comment by akvadrako** [comment]
However looking directly at the [wayland spec](https://wayland.freedesktop.org/docs/html/apa.html#protocol-spec-wl_surface), it does seem sway is more correct:

> A server should avoid signaling the frame callbacks if the surface is not visible in any way, e.g. the surface is off-screen, or completely obscured by other opaque surfaces.

**Comment by kovidgoyal** [comment]
On Tue, May 19, 2020 at 09:26:35PM -0700, Devin Bayer wrote:
> However looking directly at the [wayland spec](https://wayland.freedesktop.org/docs/html/apa.html#protocol-spec-wl_surface), it does seem sway is more correct:
> 
> > A server should avoid signaling the frame callbacks if the surface is not visible in any way, e.g. the surface is off-screen, or completely obscured by other opaque surfaces.

Yes, I'm saying it should drop the requests, not choke on them.

**Comment by akvadrako** [comment]
I don't see a mention in the protocol of the legality either way of that behavior. Is that part of the spec or how weston operates?

**Comment by akvadrako** [comment]
If Weston doesn't drop requests and the spec doesn't say to either, it doesn't seem like a sway bug. Other compositors will break too. Also, even if the compositor behaved that way, won't the client eventually crash because it is remembering all the callback IDs it's generating?

**Comment by kovidgoyal** [comment]
No

> I don't see a mention in the protocol of the legality either way of that behavior. Is that part of the spec or how weston operates?

As far as I know the spec is silent on the issue. But its pretty obvious, that the robust thing to do is either: a) respond to the latest request at the time you want the client to render or b) send all the responses at once, which even if its is for thousands of requests should take no time, since the response consists of writing a few bytes to a socket. There is a limit to how many workarounds I am willing to make for sway, just grep the kitty commit log for sway.

**Comment by akvadrako** [comment]
Are you dropping wayland support altogether or will you also consider this a bug if it occurs with weston?

**Comment by kovidgoyal** [comment]
If it occurs with non-trivial compositors other than sway, I will
re-consider.

**Comment by martinetd** [comment]
fwiw I'm not sure sway can (without a big architectural change). As far as I know it's not even reading from the client's wayland socket if the client isn't visible, so kitty can try to do whatever it wants events won't be processed (and thus not dropped either, they're probably already being dropped as duplicate as soon as they're read -- it can't bulk either, it's libwayland doing the reads/writes on the socket)

I don't know why you think something reliable like a message over unix socket needs retrying, the only way server could lose the request would be compositor dying and that'll be the least of our problems really...

@akvadrako afaik weston doesn't have a way to hide windows that would suspend events like sway does, I haven't tried gnome or KDE in ages though.

**Comment by kovidgoyal** [comment]
On Tue, May 19, 2020 at 11:41:07PM -0700, Dominique Martinet wrote:
> fwiw I'm not sure sway can (without a big architectural change). As far as I know it's not even reading from the client's wayland socket if the client isn't visible, so kitty can try to do whatever it wants events won't be processed (and thus not dropped either, they're probably already being dropped as duplicate as soon as they're read -- it can't bulk either, it's libwayland doing the reads/writes on the socket)

I dont follow. So sway doest read from the client's socket if its window
is not visible at all?? What if the client wants to do something in the
background, like say read/write the clipboard or open a new window in
response to a timed event or similar. Is it sway's position that
applications only get to interact with the windowing system if they are
visible?

And assume that is actually sway's position. In that case, when it does
get around to reading from the socket, why is it barfing if it sees a
lot of pending requests? It can simply read X KB, process all requests
discarding dupes at that point.


> 
> I don't know why you think something reliable like a message over unix socket needs retrying, the only way server could lose the request would be compositor dying and that'll be the least of our problems really...

The problem is not a message over a unix socket being lost, the problem
is a compositor dropping a render frame request because it is
busy/buggy/whatever. 

Take a step back and think about what a render frame request is. It is
a message from client to server saying, hey I am ready to render, please
let me know when I should do so. There is absolutely no reason for the
server to reply to multiple duplicate such requests, since the last one
is all that matters to the client. And even supposing, for some
inexplicable reason, the server wanted to do so,
a reply is a few bytes written to a UNIX socket, no reason for it to
fall over doing so.

The position sway seems to be taking, that a client should only ever
send a single request once it is ready to render imposes the restriction
on the entire ecosystem that *every* compositor be bug free, if any
compositor drops a render request for any reason, clients will appear to
freeze/stop rendering for random reasons. Whereas, asking a server to
simply discard redundant requests imposes no additional burden on the
server, no additional burden on the client and is more robust over the
entire ecosystem.

> 
> @akvadrako afaik weston doesn't have a way to hide windows that would suspend events like sway does, I haven't tried gnome or KDE in ages though.

Yeah, IIRC this whole render frame rigmarole was added to kitty to
workaround sway in the first place.

**Comment by martinetd** [comment]
> I dont follow. So sway doest read from the client's socket if its window
is not visible at all?? What if the client wants to do something in the
background, like say read/write the clipboard or open a new window in
response to a timed event or similar. Is it sway's position that
applications only get to interact with the windowing system if they are
visible?

On wayland, you don't get access to clipboard if you don't have focus - this isn't sway's position, that is baked in the wayland protocol itself for security reasons. Xwayland still transmits the clipboard around if you have focus on any other Xwayland window but that only works because Xwayland is a single wayland client...
You might be able to send a resize request while not focused but if you're not displayed what's the point? I'd be curious what gnome/kde does there too (if they drop frame events for non displayed windows in the first place; I think I've read kde at least does but I might be making that up...)

> bug-free compositors

I don't think totally ignoring a frame request is a probable bug to be honest, that would just break too many other clients - in general I agree with the principle of robustness but there's little point in doing something no other client does; since gtk/qt apps don't do it such bugs are bound to be picked up pretty fast.
At most the damage for one frame might be ignored maybe (kind of bugs that tend to glitch a bit but fix themselves aren't always getting all the attention they need), but no answer would make most clients just plain freeze and would get fixed right away one way or another, people don't like that :)

Well, I don't know, I've been back to X11 for a while now so it's not like I really care :P

**Comment by kovidgoyal** [comment]
On Wed, May 20, 2020 at 12:45:28AM -0700, Dominique Martinet wrote:
> > I dont follow. So sway doest read from the client's socket if its window
> is not visible at all?? What if the client wants to do something in the
> background, like say read/write the clipboard or open a new window in
> response to a timed event or similar. Is it sway's position that
> applications only get to interact with the windowing system if they are
> visible?
> 
> On wayland, you don't get access to clipboard if you don't have focus - this isn't sway's position, that is baked in the wayland protocol itself for security reasons. Xwayland still transmits the clipboard around if you have focus on any other Xwayland window but that only works because Xwayland is a single wayland client...

Oh yeah, didn't they also refuse initially to
implement middle click to paste because it's insecure  🤣

> You might be able to send a resize request while not focused but if you're not displayed what's the point? I'd be curious what gnome/kde does there too (if they drop frame events for non displayed windows in the first place; I think I've read kde at least does but I might be making that up...)

Forget about resize events, what if you want to open a new window?? Hell
suppose the user presses a key to open a new window and before the
client can respond switches workspace/minimizes the window, does that
mean on sway the new window will not be opened until the user switches
back to the original window?

Or say the client is doing some long running job, and the user happens
to switch away from it while waiting and the client wants to popup a job
details window on completion/error, on sway it cannot?

> 
> > bug-free compositors
> 
> I don't think totally ignoring a frame request is a probable bug to be honest, that would just break too many other clients - in general I agree with the principle of robustness but there's little point in doing something no other client does; since gtk/qt apps don't do it such bugs are bound to be picked up pretty fast.

Hey I hear you, like I said if any other compositors other than sway or
wlroots based ones have this issue, then I will modify kitty to do the
less robust thing to workaround it, at least on Wayland. 

> At most the damage for one frame might be ignored maybe (kind of bugs that tend to glitch a bit but fix themselves aren't always getting all the attention they need), but no answer would make most clients just plain freeze and would get fixed right away one way or another, people don't like that :)

That would happen only if the clients did not send redundant render frame
requests, are you sure they dont?

> 
> Well, I don't know, I've been back to X11 for a while now so it's not like I really care :P

Yeah me neither, I firmly expect to be in my grave before Wayland
becomes ready.

**Comment by kovidgoyal** [comment]
And actually, come to think of it, if someone can present me with some evidence that GTK and Qt actually, do this, i.e. send a single render frame request and then wait forever for a reply, then I will have kitty do the same as well.

**Comment by akvadrako** [comment]
> @akvadrako afaik weston doesn't have a way to hide windows that would suspend events like sway does, I haven't tried gnome or KDE in ages though.

It does have multiple workspaces and with a brief look at the code earlier it seems that it only handles frame callbacks when it redraws or destroys the surface.

I did a brief test and it has similar behaviour to sway. The weston log says:

> [10:28:47.375] libwayland: error in client communication (pid 215158)

and kitty aborts with the same error after getting thousands of delete_id messages.

**Comment by martinetd** [comment]
> Forget about resize events, what if you want to open a new window?? Hell
suppose the user presses a key to open a new window and before the
client can respond switches workspace/minimizes the window, does that
mean on sway the new window will not be opened until the user switches
back to the original window?
>
> Or say the client is doing some long running job, and the user happens
to switch away from it while waiting and the client wants to popup a job
details window on completion/error, on sway it cannot?

Hmm good point there, I actually tried and it works so the socket must be read somehow. I'll need to check sway/wlroots code I haven't kept up properly... Sorry for misinformation.

> And actually, come to think of it, if someone can present me with some evidence that GTK and Qt actually, do this, i.e. send a single render frame request and then wait forever for a reply, then I will have kitty do the same as well.

Sure, here's an equivalent log for termite (vte-based so gtk)... truncated but can upload full log somewhere if you want.
It's a bit noisy because gtk apparently keeps damaging the output even without getting frame events? weird. But there's no new frame callback
```
setup etc... a couple of frame events:
[383554.002]  -> wl_surface@25.frame(new id wl_callback@42)
[383554.071]  -> wl_surface@25.commit()
[383557.997] wl_display@1.delete_id(42)
[383558.062] wl_buffer@37.release()
[383558.076] wl_callback@42.done(231032722)
[383613.229]  -> wl_surface@25.attach(wl_buffer@37, 0, 0)
[383613.287]  -> wl_surface@25.set_buffer_scale(1)
[383613.304]  -> wl_surface@25.damage(0, 0, 1916, 1031)
[383613.342]  -> xdg_toplevel@31.set_min_size(9, 17)
[383613.361]  -> xdg_toplevel@31.set_max_size(0, 0)
[383613.381]  -> xdg_surface@21.set_window_geometry(0, 0, 1916, 1031)
[383613.623]  -> wl_surface@25.frame(new id wl_callback@42)
[383613.649]  -> wl_surface@25.commit()
[383617.301] wl_display@1.delete_id(42)
[383617.340] wl_buffer@37.release()
[383617.352] wl_callback@42.done(231032781)
[383638.482] wl_keyboard@30.leave(3502, wl_surface@25)
[383638.553] wl_pointer@11.leave(3507, wl_surface@25)
[383638.588]  -> wl_pointer@11.set_cursor(3497, wl_surface@19, 11, 12)
[383638.632]  -> wl_surface@19.attach(wl_buffer@41, 0, 0)
[383638.663]  -> wl_surface@19.set_buffer_scale(1)
[383638.681]  -> wl_surface@19.damage(0, 0, 24, 24)
[383638.718]  -> wl_surface@19.commit()
[383638.734] wl_pointer@11.frame()
[383638.774] xdg_toplevel@31.configure(1916, 1031, array)
[383638.808] xdg_surface@21.configure(3503)
[383638.844]  -> xdg_surface@21.ack_configure(3503)
[383657.575]  -> wl_surface@25.attach(wl_buffer@37, 0, 0)
[383657.627]  -> wl_surface@25.set_buffer_scale(1)
[383657.641]  -> wl_surface@25.damage(0, 0, 1916, 1031)
[383657.668]  -> xdg_toplevel@31.set_min_size(9, 17)
[383657.683]  -> xdg_toplevel@31.set_max_size(0, 0)
[383657.698]  -> xdg_surface@21.set_window_geometry(0, 0, 1916, 1031)
[383657.750]  -> wl_surface@25.frame(new id wl_callback@42)
[383657.767]  -> wl_surface@25.commit()
[383657.856] wl_surface@19.leave(wl_output@26)
stops being displayed here, last frame created a few lines above (callback@42)
[383657.880]  -> wl_pointer@11.set_cursor(3497, wl_surface@19, 11, 12)
[383657.907]  -> wl_surface@19.attach(wl_buffer@41, 0, 0)
[383657.927]  -> wl_surface@19.set_buffer_scale(1)
[383657.938]  -> wl_surface@19.damage(0, 0, 24, 24)
[383657.962]  -> wl_surface@19.commit()
[383679.937]  -> wl_shm@4.create_pool(new id wl_shm_pool@43, fd 22, 7901584)
[383680.008]  -> wl_shm_pool@43.create_buffer(new id wl_buffer@44, 0, 1916, 1031, 7664, 0)
[383701.490]  -> wl_surface@25.attach(wl_buffer@44, 0, 0)
[383701.541]  -> wl_surface@25.set_buffer_scale(1)
[383701.556]  -> wl_surface@25.damage(0, 0, 1916, 1031)
[383701.582]  -> xdg_toplevel@31.set_min_size(9, 17)
[383701.597]  -> xdg_toplevel@31.set_max_size(0, 0)
[383701.611]  -> xdg_surface@21.set_window_geometry(0, 0, 1916, 1031)
[383701.833] wl_buffer@37.release()
[383755.223]  -> wl_surface@25.attach(wl_buffer@44, 0, 0)
[383755.275]  -> wl_surface@25.set_buffer_scale(1)
[383755.289]  -> wl_surface@25.damage(0, 0, 1916, 1031)
[383755.316]  -> xdg_toplevel@31.set_min_size(9, 17)
[383755.332]  -> xdg_toplevel@31.set_max_size(0, 0)
[383755.347]  -> xdg_surface@21.set_window_geometry(0, 0, 1916, 1031)
[383807.307]  -> wl_surface@25.attach(wl_buffer@44, 0, 0)
[383807.370]  -> wl_surface@25.set_buffer_scale(1)
[383807.384]  -> wl_surface@25.damage(0, 0, 1916, 1031)
[383807.411]  -> xdg_toplevel@31.set_min_size(9, 17)
[383807.427]  -> xdg_toplevel@31.set_max_size(0, 0)
[383807.442]  -> xdg_surface@21.set_window_geometry(0, 0, 1916, 1031)
[383865.552]  -> wl_surface@25.attach(wl_buffer@44, 0, 0)
[383865.607]  -> wl_surface@25.set_buffer_scale(1)
[383865.620]  -> wl_surface@25.damage(0, 0, 1916, 1031)
... more damage snipped off
(only two seconds here but I've got a log with a bit over one minute hidden with no callbacks either; nothing from the compositor after that release ~15 lignes before)
[385653.813]  -> wl_surface@25.damage(0, 0, 1916, 1031)
[385653.849]  -> xdg_toplevel@31.set_min_size(9, 17)
[385653.865]  -> xdg_toplevel@31.set_max_size(0, 0)
[385653.880]  -> xdg_surface@21.set_window_geometry(0, 0, 1916, 1031)
[385717.848]  -> wl_surface@25.attach(wl_buffer@44, 0, 0)
[385717.898]  -> wl_surface@25.set_buffer_scale(1)
[385717.912]  -> wl_surface@25.damage(0, 0, 1916, 1031)
[385717.937]  -> xdg_toplevel@31.set_min_size(9, 17)
[385717.952]  -> xdg_toplevel@31.set_max_size(0, 0)
[385717.966]  -> xdg_surface@21.set_window_geometry(0, 0, 1916, 1031)
[385718.181] wl_display@1.delete_id(42)
^ hmm that one is curious but it appears to work even if it's used later...
[385718.203] wl_keyboard@30.enter(3520, wl_surface@25, array)
[385718.233] wl_keyboard@30.modifiers(3521, 8, 0, 0, 0)
[385718.280] wl_data_device@18.selection(nil)
[385718.301] gtk_primary_selection_device@17.data_offer(new id gtk_primary_selection_offer@20623616)
[385718.316] gtk_primary_selection_offer@4278190080.offer("text/plain;charset=utf-8")
[385718.328] gtk_primary_selection_device@17.selection(gtk_primary_selection_offer@4278190080)
[385718.341]  -> gtk_primary_selection_offer@4278190081.destroy()
[385718.356] wl_pointer@11.enter(3523, wl_surface@25, 603.394531, 341.476562)
[385718.389]  -> wl_pointer@11.set_cursor(3523, wl_surface@19, 11, 12)
[385718.414]  -> wl_surface@19.attach(wl_buffer@41, 0, 0)
[385718.433]  -> wl_surface@19.set_buffer_scale(1)
[385718.444]  -> wl_surface@19.damage(0, 0, 24, 24)
[385718.466]  -> wl_surface@19.commit()
[385718.475] wl_pointer@11.frame()
[385718.515] xdg_toplevel@31.configure(1916, 1031, array)
[385718.547] xdg_surface@21.configure(3519)
[385718.575]  -> xdg_surface@21.ack_configure(3519)
earlier frame callback answered here:
[385718.588] wl_callback@42.done(231034859)
[385731.958]  -> wl_surface@25.attach(wl_buffer@44, 0, 0)
[385732.009]  -> wl_surface@25.set_buffer_scale(1)
[385732.036]  -> wl_surface@25.damage(0, 0, 1916, 1031)
[385732.064]  -> xdg_toplevel@31.set_min_size(9, 17)
[385732.078]  -> xdg_toplevel@31.set_max_size(0, 0)
[385732.092]  -> xdg_surface@21.set_window_geometry(0, 0, 1916, 1031)
[385732.142]  -> wl_surface@25.frame(new id wl_callback@42)
[385732.165]  -> wl_buffer@37.destroy()
and back to business as usual ^
```

**Comment by martinetd** [comment]
Well, ok, I was slow. Thanks anyway :)

**Comment by travankor** [comment]
>Hey I hear you, like I said if any other compositors other than sway or
wlroots based ones have this issue, then I will modify kitty to do the
less robust thing to workaround it, at least on Wayland.

There is also https://github.com/Smithay/smithay/ but it's not ready yet. Unfortunately, power users don't have much choice yet on Wayland.

**Comment by rawoul** [comment]
Hi all,
I wonder if the problem might actually be that the main display wayland event queue is not being polled regularly and dispatched properly by the client (either kitty or glfw, I'm not sure of this is handled in kitty), which after some time would result in the unix socket buffer being full. This in turn will result in an error in the compositor as it cannot push events anymore, which is a fatal error resulting in the socket being closed.

This can happen when creating a separate wl_event_queue in wayland to allow processing events in a separate thread. In this case there might still be some events sent on the main wl_display event queue, like the delete_id events we see here. The main wl_display event_queue must also be polled on dispatched regularly.

Just my two cents, I'll see if I have some time to look at the code and check if that could be the issue.

**Comment by kovidgoyal** [comment]
This was already fixed two weeks ago, see the closing commit.

**Comment by rawoul** [comment]
> This was already fixed two weeks ago, see the closing commit.

I saw that, what I'm saying is that the underlying issue might still be there, and the crash could still happen. The current fix of throttling rendering does prevent generating many delete_id events, but if there's another reason where the compositor might be queueing many events and kitty does not dispatch them, the crash will still happen. Wayland clients or compositors should not have any problem rendering offscreen at any rate.

**Comment by kovidgoyal** [comment]
The underlying issue is that sway falls over and dies if it gets too many render frame requests that it is not willing to service. kitty doesn't send those anymore, so the underlying issue is fixed. And no kitty has absolutely no problem dispatching all the events the compositor might want to send it. And there is no render throttling here except that imposed by the compositor, which kitty can do nothing about. The compositor is the one that is closing the pipe and not because it is full.

**Comment by jnpkrn** [comment]
The patch does help for the problem in question at my end, thanks.

**Comment by jonjomckay** [comment]
I'm still running into this using v0.18.1, from Nix running under Arch Linux. I can't reliably reproduce it, but it does somehow break my GPU, causing everything to crash, then things render really slowly until I reboot. Is there anything I can do to help diagnose the problem further?

**Comment by martinetd** [comment]
with that kind of symptoms it's more likely to be a GPU bug, probably unrelated and not much kitty can do about it.
check dmesg etc there might be some other hint

**Comment by kovidgoyal** [comment]
That does not sound anything like this bug, this bug was simply sway not
being able to handle many unfulfilled render frame requests, and so closing its
socket to kitty. There's no way it could possibly cause your GPU to die,
since it generates no activity on the GPU whatsoever.

**Comment by jonjomckay** [comment]
I'm pretty sure it sounds very similar to this bug, as I get the same message, and it also happens when tailing the logs for an application that logs a lot of stuff quickly, and it only manifested since I swapped from i3 (Xorg) to Sway (Wayland). I was using Kitty with no issues up until then, with the exact same workflow, same GPU and same driver (Nouveau).

The `[glfw error 65544]: Wayland: fatal display error: Broken pipe` message I also get suggests to me that it is this bug that is causing my GPU to go haywire, especially as I experience no graphical issues until this bug occurs - Sway suddenly closes and switches me back to the TTY, where that log message is echoing constantly, and the message continues until I manually `kill` all `kitty` processes, stopping the messages.

Once I've killed the processes, and restart Sway, my screens flicker constantly, all Kitty terminal output renders very slowly, and sometimes random triangles display followed by the system completely locking up.

@martinetd I've checked `dmesg` but there's no helpful output there - is there anywhere else you recommend I look, or anything else I can do to help debug this?

**Comment by kovidgoyal** [comment]
If sway is crashing then that will generate broken pipe messages, since a crash will break the pipe, which have nothing to do with this bug. And if sway is crashing you should report the bug to sway. Again, this issue causes absolutely no traffic to the GPU so it cannot affect GPU state.

**Comment by Poly2it** [comment]
> I'm pretty sure it sounds very similar to this bug, as I get the same message, and it also happens when tailing the logs for an application that logs a lot of stuff quickly, and it only manifested since I swapped from i3 (Xorg) to Sway (Wayland). I was using Kitty with no issues up until then, with the exact same workflow, same GPU and same driver (Nouveau).
> 
> The `[glfw error 65544]: Wayland: fatal display error: Broken pipe` message I also get suggests to me that it is this bug that is causing my GPU to go haywire, especially as I experience no graphical issues until this bug occurs - Sway suddenly closes and switches me back to the TTY, where that log message is echoing constantly, and the message continues until I manually `kill` all `kitty` processes, stopping the messages.
> 
> Once I've killed the processes, and restart Sway, my screens flicker constantly, all Kitty terminal output renders very slowly, and sometimes random triangles display followed by the system completely locking up.
> 
> @martinetd I've checked `dmesg` but there's no helpful output there - is there anywhere else you recommend I look, or anything else I can do to help debug this?

Were you able to solve this? I am experiencing the same issue on Hyprland. Very annoying since it basically forces me to restart my computer and kills whatever I was working with.

**Comment by Poly2it** [comment]
@jonjomckay

**Comment by francoposa** [comment]
this is now occurring on vanilla Gnome on Fedora as of a few months ago - results in a crash of Gnome which manifests as being logged out of the session.
